src/redis/listpack.c

```diff
@@ -649,6 +649,75 @@ lpGetWithSize(unsigned char *p, int64_t *count, unsigned char *intbuf, uint64_t
     }
 }
 
+unsigned char *lpGet2(unsigned char *p, int64_t *count) {
+    int64_t val;
+    uint64_t uval = 0, negstart = UINT64_MAX, negmax = 0;
+    uint8_t encoding = p[0];
+    
+    // Prioritize checking for integers first.
+    if (encoding < LP_ENCODING_7BIT_UINT_MASK) {        
+        uval = encoding & 0x7f;    
+    } else if (encoding > LP_ENCODING_32BIT_STR) {
+        switch (encoding) {
+            case LP_ENCODING_16BIT_INT:
+                uval = (uint64_t)p[1] | (uint64_t)p[2] << 8;
+                negstart = (uint64_t)1<<15;
+                negmax = UINT16_MAX;
+                break;
+            case LP_ENCODING_24BIT_INT:
+                uval = (uint64_t)p[1] | (uint64_t)p[2] << 8 | (uint64_t)p[3] << 16;
+                negstart = (uint64_t)1<<23;
+                negmax = UINT32_MAX>>8;
+                break;
+            case LP_ENCODING_32BIT_INT:
+                uval = (uint64_t)p[1] | (uint64_t)p[2] << 8 | (uint64_t)p[3] << 16 | (uint64_t)p[4] << 24;
+                negstart = (uint64_t)1<<31;
+                negmax = UINT32_MAX;
+                break;
+            case LP_ENCODING_64BIT_INT:                
+                uval = (uint64_t)p[1] | (uint64_t)p[2] << 8 | (uint64_t)p[3] << 16 | (uint64_t)p[4] << 24 |
+               (uint64_t)p[5] << 32 | (uint64_t)p[6] << 40 | (uint64_t)p[7] << 48 | (uint64_t)p[8] << 56;
+                negstart = (uint64_t)1<<63;
+                negmax = UINT64_MAX;
+            break;
+            default:
+                // Invalid encoding, return a large integer value.
+                uval = 12345678900000000ULL + p[0];
```

Comment:
I think this part can be vectorized using the example x86  _mm_slli_epi16 instruction (maybe exist some extension for GCC or we can use separate intrinsic for arm and x86) also exist vector instructions that merge two int8 into int16 and 2 int16 into int32

Reply:
I do not see what can be vectorized here. 
The whole logic is around a single byte

Reply:
I select too much,   all shifts and or operations like
 uval = (uint64_t)p[1] | (uint64_t)p[2] << 8 | (uint64_t)p[3] << 16 | (uint64_t)p[4] << 24 |
               (uint64_t)p[5] << 32 | (uint64_t)p[6] << 40 | (uint64_t)p[7] << 48 | (uint64_t)p[8] << 56;
                negstart = (uint64_t)1<<63;
                negmax = UINT64_MAX;

Reply:
if you paste it into godbolt you can see it has been compiled into a single load instruction. in any case, this is why I added micro benchmarks - to be able to iterate and improve if needed. I think it's fairly good as is.